core/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -91,6 +94,36 @@ private[spark] abstract class YarnSchedulerBackend(
   }
 
   /**
+   * Override the DriverEndpoint to add extra logic for the case when an executor is disconnected.
+   * We should check the cluster manager and find if the loss of the executor was caused by YARN
+   * force killing it due to preemption.
+   */
+  private class YarnDriverEndpoint(rpcEnv: RpcEnv, sparkProperties: Seq[(String, String)])
+      extends DriverEndpoint(rpcEnv, sparkProperties) {
```

Comment:
I don't see the benefit in keeping them separate. Now the developer has to worry about which one to send a message to. If s/he accidentally picks the wrong one then we'll just time out on a future, which is hard to debug. Also they're not actually totally separate, since one calls into another.

Anyway I don't feel strongly about this because keeping them separate is actually just preserving existing behavior. I just find the introduction of another endpoint in the same file a little unnecessary.